File: leetcode/dp/647. Palindromic Substrings/py/main.py
class Solution:
    # Time O(N^2)
    # Space O(N^2)
    def countSubstrings(self, s: str) -> int:
        n = len(s)
        dp = [[False] * n for _ in range(n)]
        ans = 0

        for i in range(n):
            dp[i][i] = True

        for curLen in range(1, n):
          for i in range(n - curLen):
              j = i + curLen

              if s[i] == s[j]:
                  dp[i][j] = True if j - i == 1 else dp[i + 1][j - 1]
              
              if dp[i][j]:
                  ans += 1

        return ans + n

File: leetcode/dp/647. Palindromic Substrings/js/main.js
/*

State
  state[i][j] is true if substring s[i, j] is palindromic

Aim State
  decide all possible state[i][j] and count the element that is true

State Transition
  state[i][j] is true if s[i] == s[j] and state[i+1][j-1] is true (j - i >= 2)
  state[i][j] is true if s[i] == s[j] (j - i == 1)
  state[i][j] is true (j - i == 0)

*/

// Time O(N^2)
// Space O(N^2)
const countSubstrings = s => {
  let n = s.length;
  let ans = 0;
  let dp = Array(n)
    .fill(null)
    .map(() => Array(n).fill(false));

  for (let i = n - 1; i >= 0; i--) {
    for (let j = i; j < n; j++) {
      if (s[i] === s[j] && (j - i < 3 || dp[i + 1][j - 1])) {
        dp[i][j] = true;
      }

      if (dp[i][j]) {
        ans++;
      }
    }
  }

  return ans;
};
